src/itwinai/torch/trainer.py

```diff
@@ -1220,3 +1231,390 @@ def dist_train(
+                pin_memory=pin_memory,
+                generator=self.torch_rng,
+                shuffle=shuffle_test,
+                sampler=sampler,
+                collate_fn=collate_fn,
+            )
+        else:
+            self.test_dataloader = None
+
+    @abstractmethod
+    def train(config, data=None):
+        pass
+
+    @monitor_exec
+    def execute(
+        self,
+        train_dataset: Dataset,
+        validation_dataset: Optional[Dataset] = None,
+        test_dataset: Optional[Dataset] = None,
+    ) -> Tuple[Dataset, Dataset, Dataset, Any]:
+        """Execute the training pipeline with the given datasets.
+
+        Args:
+            train_dataset (Dataset): Training dataset.
+            validation_dataset (Optional[Dataset], optional): Validation dataset.
+                Defaults to None.
+            test_dataset (Optional[Dataset], optional): Test dataset. Defaults to None.
+
+        Returns:
+            Tuple[Dataset, Dataset, Dataset, Any]:
+                A tuple containing the datasets and the training result grid.
+        """
+        train_with_data = tune.with_parameters(
+            self.train, data=[train_dataset, validation_dataset, test_dataset]
+        )
+        trainer = ray.train.torch.TorchTrainer(
+            train_with_data,
+            train_loop_config=self.train_loop_config,
+            scaling_config=self.scaling_config,
+            run_config=self.run_config,
+        )
+        param_space = {"train_loop_config": self.train_loop_config}
+        tuner = tune.Tuner(trainer, param_space=param_space, tune_config=self.tune_config)
+
+        result_grid = tuner.fit()
+
+        return train_dataset, validation_dataset, test_dataset, result_grid
+
+    def set_epoch(self, epoch: int) -> None:
+        self.train_dataloader.sampler.set_epoch(epoch)
+        if self.validation_dataloader is not None:
+            self.validation_dataloader.sampler.set_epoch(epoch)
+        if self.test_dataloader is not None:
+            self.test_dataloader.sampler.set_epoch(epoch)
+
+    def _set_tune_config(self):
+        tune_config = self.config.get("tune_config")
+        if tune_config:
+            num_samples = tune_config.get("num_samples", 10)
+            metric = tune_config.get("metric", "loss")
+            mode = tune_config.get("mode", "min")
+
+            search_alg = (
+                get_raytune_search_alg(tune_config) if "search_alg" in tune_config else None
+            )
+            scheduler = (
+                get_raytune_schedule(tune_config) if "scheduler" in tune_config else None
+            )
+
+            # Only set metric and mode if search_alg and scheduler aren't defined
+            self.tune_config = tune.TuneConfig(
+                num_samples=num_samples,
+                metric=metric if not search_alg and not scheduler else None,
+                mode=mode if not search_alg and not scheduler else None,
```

Comment:
Personally I think these are a little too convoluted to do inline. I would make variables before `self.tune_config = ...`